packages/stargate/src/signingstargateclient.spec.ts

```diff
@@ -0,0 +1,168 @@
+import { coin, coins, GasPrice } from "@cosmjs/launchpad";
+import { DirectSecp256k1Wallet, Registry } from "@cosmjs/proto-signing";
+import { assert } from "@cosmjs/utils";
+
+import { cosmos } from "./codec";
+import { PrivateSigningStargateClient, SigningStargateClient } from "./signingstargateclient";
+import { assertIsBroadcastTxSuccess } from "./stargateclient";
+import { faucet, makeRandomAddress, pendingWithoutSimapp, simapp, validator } from "./testutils.spec";
+
+describe("SigningStargateClient", () => {
+  describe("constructor", () => {
+    it("can be constructed with default fees", async () => {
+      const wallet = await DirectSecp256k1Wallet.fromMnemonic(faucet.mnemonic);
+      const client = await SigningStargateClient.connectWithWallet(simapp.tendermintUrl, wallet);
+      const openedClient = (client as unknown) as PrivateSigningStargateClient;
+      expect(openedClient.fees).toEqual({
+        send: {
+          amount: [
+            {
+              amount: "2000",
+              denom: "ucosm",
+            },
+          ],
+          gas: "80000",
+        },
+      });
+    });
+
+    it("can be constructed with custom registry", async () => {
+      const wallet = await DirectSecp256k1Wallet.fromMnemonic(faucet.mnemonic);
+      const registry = new Registry();
+      registry.register("/custom.MsgCustom", cosmos.bank.v1beta1.MsgSend);
+      const client = await SigningStargateClient.connectWithWallet(simapp.tendermintUrl, wallet, registry);
+      const openedClient = (client as unknown) as PrivateSigningStargateClient;
+      expect(openedClient.registry.lookupType("/custom.MsgCustom")).toEqual(cosmos.bank.v1beta1.MsgSend);
+    });
+
+    it("can be constructed with custom gas price", async () => {
+      const wallet = await DirectSecp256k1Wallet.fromMnemonic(faucet.mnemonic);
+      const gasPrice = GasPrice.fromString("3.14utest");
+      const client = await SigningStargateClient.connectWithWallet(
+        simapp.tendermintUrl,
+        wallet,
+        undefined,
+        gasPrice,
+      );
+      const openedClient = (client as unknown) as PrivateSigningStargateClient;
+      expect(openedClient.fees).toEqual({
+        send: {
+          amount: [
+            {
+              amount: "251200", // 3.14 * 80_000
+              denom: "utest",
+            },
+          ],
+          gas: "80000",
+        },
+      });
+    });
+
+    it("can be constructed with custom gas limits", async () => {
+      const wallet = await DirectSecp256k1Wallet.fromMnemonic(faucet.mnemonic);
+      const gasLimits = {
+        send: 160000,
+      };
+      const client = await SigningStargateClient.connectWithWallet(
+        simapp.tendermintUrl,
+        wallet,
+        undefined,
```

Comment:
These are lots of optional fields. I am happy with one `option?: T` at the end, which may be there. With two, where you need `(x, undefined, y)` it starts to get ugly. With 3 here even more.

Why not do `connectWithWallet(url, wallet, options?: Options)`, where

```ts
interface Options {
   registry: Registry,
   gasPrice: string
   gasLimits: T
}
```

usage would be simple as well, and much clearer:

`await SigningStargateClient.connectWithWallet(simapp.tendermintUrl, wallet, {gasLimits});`